pkg/operator/csr/cert_controller.go

```diff
@@ -0,0 +1,322 @@
+		c.reset()
+		return nil
+	}
+
+	// create a csr to request new client certificate if
+	// a. there is no valid client certificate issued for the current cluster/agent
+	// b. client certificate exists and has less than a random percentage range from 20% to 25% of its life remaining
+	if c.hasValidClientCertificate(secret) {
+		notBefore, notAfter, err := getCertValidityPeriod(secret)
+		if err != nil {
+			return err
+		}
+
+		total := notAfter.Sub(*notBefore)
+		remaining := notAfter.Sub(time.Now())
+		klog.V(4).Infof("Client certificate for %s: time total=%v, remaining=%v, remaining/total=%v", c.controllerName, total, remaining, remaining.Seconds()/total.Seconds())
+		threshold := jitter(0.2, 0.25)
+		if remaining.Seconds()/total.Seconds() > threshold {
+			// Do nothing if the client certificate is valid and has more than a random percentage range from 20% to 25% of its life remaining
+			klog.V(4).Infof("Client certificate for %s is valid and has more than %.2f%% of its life remaining", c.controllerName, threshold*100)
+			return nil
+		}
+		syncCtx.Recorder().Eventf("CertificateRotationStarted", "The current client certificate for %s expires in %v. Start certificate rotation", c.controllerName, remaining.Round(time.Second))
+	} else {
+		syncCtx.Recorder().Eventf("NoValidCertificateFound", "No valid client certificate for %s is found. Bootstrap is required", c.controllerName)
+	}
+
+	// create a new private key
+	c.keyData, err = keyutil.MakeEllipticPrivateKeyPEM()
+	if err != nil {
+		return err
+	}
+
+	// create a csr
+	c.csrName, err = c.createCSR(ctx)
+	if err != nil {
+		c.reset()
+		return err
+	}
+	syncCtx.Recorder().Eventf("CSRCreated", "A csr %q is created for %s", c.csrName, c.controllerName)
+	return nil
+}
+
+func (c *clientCertificateController) syncCSR(secret *corev1.Secret) (map[string][]byte, error) {
+	// skip if there is no ongoing csr
+	if len(c.csrName) == 0 {
+		return nil, fmt.Errorf("no ongoing csr")
+	}
+
+	// skip if csr no longer exists
+	csr, err := c.hubCSRLister.Get(c.csrName)
+	switch {
+	case errors.IsNotFound(err):
+		// fallback to fetching csr from hub apiserver in case it is not cached by informer yet
+		csr, err = c.hubCSRClient.Get(context.Background(), c.csrName, metav1.GetOptions{})
+		if errors.IsNotFound(err) {
+			return nil, fmt.Errorf("unable to get csr %q. It might have already been deleted.", c.csrName)
+		}
+	case err != nil:
+		return nil, err
+	}
+
+	// skip if csr is not approved yet
+	if !isCSRApproved(csr) {
+		return nil, nil
+	}
+
+	// skip if csr has no certificate in its status yet
+	if len(csr.Status.Certificate) == 0 {
+		return nil, nil
+	}
+
+	klog.V(4).Infof("Sync csr %v", c.csrName)
+	// check if cert in csr status matches with the corresponding private key
+	if c.keyData == nil {
+		return nil, fmt.Errorf("No private key found for certificate in csr: %s", c.csrName)
+	}
+	_, err = tls.X509KeyPair(csr.Status.Certificate, c.keyData)
```

Comment:
the docs do not mention the checks `tls.X509KeyPair` does, I would rather see this done expanded